parser/jsonquery_visitor_impl.go

```diff
@@ -116,6 +118,65 @@ func (j *JsonQueryVisitorImpl) VisitPresentExp(ctx *PresentExpContext) interface
 	return j.leftOp != nil
 }
 
+func (j *JsonQueryVisitorImpl) VisitCompareVersionExp(ctx *CompareVersionExpContext) interface{} {
+	ctx.AttrPath().Accept(j)
+	ctx.Value().Accept(j)
+	if j.hasErr() {
+		return false
+	}
+	var apply func(Operand, Operand) (bool, error)
+
+	// force to use version operation
+	currentOp := &VersionOperation{}
+	switch ctx.op.GetTokenType() {
+	case JsonQueryParserEQ:
+		apply = currentOp.EQ
+	case JsonQueryParserNE:
+		apply = currentOp.NE
+	case JsonQueryParserGT:
+		apply = currentOp.GT
+	case JsonQueryParserLT:
+		apply = currentOp.LT
+	case JsonQueryParserLE:
+		apply = currentOp.LE
+	case JsonQueryParserGE:
+		apply = currentOp.GE
+	default:
+		j.setErr(fmt.Errorf("Unknown operation %s", ctx.op.GetText()))
+		return false
+	}
+
+	// force double and int into version. workaround for same lexer and identical parser but different logic
+	// solve x.y and x as token version instead of double / int, prevent mess on lexer tokenization
+	switch j.rightOp.(type) {
+	case float64:
+		value := j.rightOp.(float64)
+		if value < 0 {
+			return false
+		}
+		j.rightOp = strconv.FormatFloat(value, 'f', -1, 64)
+		break
+	case int:
+		value := j.rightOp.(int)
+		if value < 0 {
+			return false
+		}
+		j.rightOp = strconv.FormatInt(int64(value), 10)
+		break
+	case string:
+		break
```

Comment:
ini kalo case nya `1.0.0-alpha` jadi bisa di compare ndak ya?

Reply:
interesting cases 👍 

is `1.0.0-alpha` on payload (`map[string]interface{}`), or on the rule (`compareVersion(appVersion > 1.0.0-alpha)`)?

from lexer rules, it should not being recognized. will update it to recognize it.

i think if it is from payload, it should still can be compared

Reply:
from rules mas @endrawanandika, like this one `compareVersion(appVersion > 1.0.0-alpha)`

Reply:
it is possible with `compareVersion(appVersion > "1.0.0-alpha")`. i already add test cases regarding that cases.

also i add non semantic version `1.0.0-dev` on payload to satisfy our use case, as we discussed last time